osu.Game/Online/Multiplayer/StatefulMultiplayerClient.cs

```diff
@@ -109,30 +108,66 @@ protected StatefulMultiplayerClient()
             });
         }
 
+        private readonly object joinOrLeaveTaskLock = new object();
+        private Task? joinOrLeaveTask;
+
         /// <summary>
         /// Joins the <see cref="MultiplayerRoom"/> for a given API <see cref="Room"/>.
         /// </summary>
         /// <param name="room">The API <see cref="Room"/>.</param>
         public async Task JoinRoom(Room room)
         {
-            if (Room != null)
-                throw new InvalidOperationException("Cannot join a multiplayer room while already in one.");
+            Task? lastTask;
+            Task newTask;
 
-            Debug.Assert(room.RoomID.Value != null);
+            lock (joinOrLeaveTaskLock)
+            {
+                lastTask = joinOrLeaveTask;
+                joinOrLeaveTask = newTask = Task.Run(async () =>
+                {
+                    if (lastTask != null)
+                        await lastTask;
 
-            apiRoom = room;
-            playlistItemId = room.Playlist.SingleOrDefault()?.ID ?? 0;
+                    // Should be thread-safe since joinOrLeaveTask is locked on in both JoinRoom() and LeaveRoom().
+                    if (Room != null)
+                        throw new InvalidOperationException("Cannot join a multiplayer room while already in one.");
 
-            Room = await JoinRoom(room.RoomID.Value.Value);
+                    Debug.Assert(room.RoomID.Value != null);
 
-            Debug.Assert(Room != null);
+                    // Join the server-side room.
+                    var joinedRoom = await JoinRoom(room.RoomID.Value.Value);
+                    Debug.Assert(joinedRoom != null);
 
-            var users = await getRoomUsers();
-            Debug.Assert(users != null);
+                    // Populate users.
+                    Debug.Assert(joinedRoom.Users != null);
+                    await Task.WhenAll(joinedRoom.Users.Select(PopulateUser));
 
-            await Task.WhenAll(users.Select(PopulateUser));
+                    // Update the stored room (must be done on update thread for thread-safety).
+                    await scheduleAsync(() =>
+                    {
+                        Room = joinedRoom;
+                        apiRoom = room;
+                        playlistItemId = room.Playlist.SingleOrDefault()?.ID ?? 0;
+                    });
+
+                    // Update room settings.
+                    await updateLocalRoomSettings(joinedRoom.Settings);
+                });
+            }
 
-            updateLocalRoomSettings(Room.Settings);
+            try
+            {
+                await newTask;
+            }
+            finally
+            {
+                // The task will be awaited in the future, so reset it so that the user doesn't get into a permanently faulted state if anything fails.
+                lock (joinOrLeaveTask)
```

Comment:
Should take
```suggestion
                lock (joinOrLeaveTaskLock)
```
right?